pkg/executor/sortexec/sort_partition.go

```diff
@@ -0,0 +1,310 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sortexec
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/failpoint"
+	"github.com/pingcap/tidb/pkg/types"
+	"github.com/pingcap/tidb/pkg/util/chunk"
+	"github.com/pingcap/tidb/pkg/util/disk"
+	"github.com/pingcap/tidb/pkg/util/memory"
+)
+
+var errSpillEmptyChunk = errors.New("can not spill empty chunk to disk")
+var errFailToAddChunk = errors.New("fail to add chunk")
+
+// It should be const, but we need to modify it for test.
+var spillChunkSize int = 4096
+
+// signalCheckpointForSort indicates the times of row comparation that a signal detection will be triggered.
+const signalCheckpointForSort uint = 10240
+
+type sortPartition struct {
+	fieldTypes []*types.FieldType
+
+	// Store data here
+	savedRows          []chunk.Row
+	totalTrackedMemNum int64
+	isSorted           bool
+
+	inDisk *chunk.DataInDiskByChunks
+
+	memTracker  *memory.Tracker
+	diskTracker *disk.Tracker
+	spillAction *sortPartitionSpillDiskAction
+	helper      *spillHelper
+
+	// We can't spill if size of data is lower than the limit
+	spillLimit int64
+
+	byItemsDesc []bool
+	// keyColumns is the column index of the by items.
+	keyColumns []int
+	// keyCmpFuncs is used to compare each ByItem.
+	keyCmpFuncs []chunk.CompareFunc
+
+	// Sort is a time-consuming operation, we need to set a checkpoint to detect
+	// the outside signal periodically.
+	timesOfRowCompare uint
+
+	spillTriggered bool
+
+	// It's index that points to the data need to be read next time.
+	idx int
+}
+
+// Creates a new SortPartition in memory.
+func newSortPartition(fieldTypes []*types.FieldType, chunkSize int, byItemsDesc []bool,
+	keyColumns []int, keyCmpFuncs []chunk.CompareFunc, helper *spillHelper, spillLimit int64) *sortPartition {
+	helper.reset()
+	retVal := &sortPartition{
+		fieldTypes:         fieldTypes,
+		savedRows:          make([]chunk.Row, 0),
+		totalTrackedMemNum: 0,
+		isSorted:           false,
+		inDisk:             nil, // It's initialized only when spill is triggered
+		memTracker:         memory.NewTracker(memory.LabelForSortPartition, -1),
+		diskTracker:        disk.NewTracker(memory.LabelForSortPartition, -1),
+		spillAction:        nil, // It's set in `actionSpill` function
+		helper:             helper,
+		spillLimit:         spillLimit,
+		byItemsDesc:        byItemsDesc,
+		keyColumns:         keyColumns,
+		keyCmpFuncs:        keyCmpFuncs,
+		spillTriggered:     false,
+		idx:                0,
+	}
+
+	return retVal
+}
+
+func (s *sortPartition) close() error {
+	if s.inDisk != nil {
+		s.inDisk.Close()
+		s.inDisk = nil
+	}
+	s.getMemTracker().UnconsumeAll()
```

Comment:
s.getMemTracker().ReplaceBytesUsed(0) is ok.

Reply:
> s.getMemTracker().ReplaceBytesUsed(0) is ok.

okk, I will fix it.